// @udex/web-components/dist/Carousel.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var Carousel_1;
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import slot from "@ui5/webcomponents-base/dist/decorators/slot.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import { getI18nBundle } from "@ui5/webcomponents-base/dist/i18nBundle.js";
import { isUp, isDown, isRight, isLeft, } from "@ui5/webcomponents-base/dist/Keys.js";
import UI5Carousel from "@ui5/webcomponents/dist/Carousel.js";
import { CAROUSEL_NEXT_ARROW_TEXT, CAROUSEL_PREVIOUS_ARROW_TEXT, } from "./generated/i18n/i18n-defaults.js";
import redirectAttributes from "./utils/redirectAttributes.js";
import CarouselTemplate from "./generated/templates/CarouselTemplate.lit.js";
// Styles
import CarouselCss from "./generated/themes/Carousel.css.js";
export var CarouselNavigationPosition;
(function (CarouselNavigationPosition) {
    CarouselNavigationPosition["top"] = "top";
    CarouselNavigationPosition["bottom"] = "bottom";
})(CarouselNavigationPosition || (CarouselNavigationPosition = {}));
/**
 * @class
 *
 * @constructor
 * @alias udex.components.Carousel
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-carousel
 * @public
 */
let Carousel = Carousel_1 = class Carousel extends UI5Element {
    static async onDefine() {
        Carousel_1.i18nBundle = await getI18nBundle("sap-ui-webcomponents-bundle");
    }
    get nextPageText() {
        return Carousel_1.i18nBundle.getText(CAROUSEL_NEXT_ARROW_TEXT);
    }
    get previousPageText() {
        return Carousel_1.i18nBundle.getText(CAROUSEL_PREVIOUS_ARROW_TEXT);
    }
    get carouselClass() {
        const classes = "udex-carousel";
        return `${classes} udex-carousel__navigation--${this.navigationPlacement}`;
    }
    get buttonDesign() {
        return this.navigationPlacement === CarouselNavigationPosition.top
            ? "Secondary"
            : "Tertiary";
    }
    get showNavigationButtons() {
        return {
            top: this.navigationPlacement === CarouselNavigationPosition.top,
            bottom: this.navigationPlacement === CarouselNavigationPosition.bottom,
        };
    }
    get _hidePageIndicator() {
        return this.navigationPlacement === CarouselNavigationPosition.top;
    }
    get _carousel() {
        return this.shadowRoot.querySelector("[ui5-carousel]");
    }
    get _ui5CarouselRoot() {
        return this._carousel?.shadowRoot?.querySelector(".ui5-carousel-root");
    }
    _calculateDisabledButtonsState() {
        this.isPrevButtonDisabled = !this._carousel?.hasPrev;
        this.isNextButtonDisabled = !this._carousel?.hasNext;
    }
    _prevButtonClick() {
        if (this.isPrevButtonDisabled) {
            return;
        }
        this._carousel?.navigateLeft();
        this._ui5CarouselRoot?.focus();
        this._calculateDisabledButtonsState();
    }
    _nextButtonClick() {
        if (this.isNextButtonDisabled) {
            return;
        }
        this._carousel?.navigateRight();
        this._ui5CarouselRoot?.focus();
        this._calculateDisabledButtonsState();
    }
    _onkeydown(e) {
        if (isLeft(e) || isDown(e) || isRight(e) || isUp(e)) {
            this._calculateDisabledButtonsState();
        }
    }
    onAfterRendering() {
        redirectAttributes(this, "carousel");
        this._calculateDisabledButtonsState();
    }
};
__decorate([
    property({
        type: CarouselNavigationPosition,
        defaultValue: CarouselNavigationPosition.top,
    })
], Carousel.prototype, "navigationPlacement", void 0);
__decorate([
    slot({ "default": true, type: HTMLElement, individualSlots: true })
], Carousel.prototype, "items", void 0);
__decorate([
    property({ type: Boolean, noAttribute: true })
], Carousel.prototype, "isPrevButtonDisabled", void 0);
__decorate([
    property({ type: Boolean, noAttribute: true })
], Carousel.prototype, "isNextButtonDisabled", void 0);
Carousel = Carousel_1 = __decorate([
    customElement({
        tag: "udex-carousel",
        renderer: litRender,
        styles: CarouselCss,
        template: CarouselTemplate,
        dependencies: [UI5Carousel],
    })
], Carousel);
Carousel.define();
export default Carousel;
