// @udex/web-components/dist/Checkbox.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import "@ui5/webcomponents/dist/CheckBox.js";
import CheckboxTemplate from "./generated/templates/CheckboxTemplate.lit.js";
// Styles
import CheckboxCss from "./generated/themes/Checkbox.css.js";
import redirectAttributes from "./utils/redirectAttributes.js";
/**
 * @class
 *
 * <h3 class="comment-api-title">Overview</h3>
 *
 *
 * <h3>Usage</h3>
 *
 * For the <code>udex-checkbox</code>
 * <h3>ES6 Module Import</h3>
 *
 * <code>import @udex/web-components/dist/Checkbox.js";</code>
 *
 * @constructor
 * @author SAP SE
 * @alias sap.ui.webc.web-components.Checkbox
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-checkbox
 * @public
 */
let Checkbox = class Checkbox extends UI5Element {
    onAfterRendering() {
        redirectAttributes(this, "udex-checkbox");
    }
};
Checkbox = __decorate([
    customElement({
        tag: "udex-checkbox",
        renderer: litRender,
        styles: CheckboxCss,
        template: CheckboxTemplate,
        dependencies: [],
    })
], Checkbox);
Checkbox.define();
export default Checkbox;
