// @udex/web-components/dist/Footer.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import "@ui5/webcomponents-icons/dist/headset.js";
import "@ui5/webcomponents-icons/dist/email.js";
import "@ui5/webcomponents-icons/dist/back-to-top.js";
import "@ui5/webcomponents-icons/dist/discussion-2.js";
import "@ui5/webcomponents-icons/dist/navigation-right-arrow.js";
// Template
import FooterTemplate from "./generated/templates/FooterTemplate.lit.js";
// Styles
import FooterCss from "./generated/themes/Footer.css";
let Footer = class Footer extends UI5Element {
    onBackToTopClick() {
        window.scrollTo(0, 0);
    }
    expandLinksList(event) {
        let currentElement = event.target;
        while (currentElement
            && !currentElement.classList.contains("udex-footer__links-group-container")) {
            currentElement = currentElement.parentElement;
        }
        if (currentElement) {
            currentElement.classList.toggle("active");
        }
    }
    parseArrayFromString(string) {
        return string ? JSON.parse(string) : [];
    }
    get quickLinksArray() {
        return this.parseArrayFromString(this.quickLinks);
    }
    get trendingLinksArray() {
        return this.parseArrayFromString(this.trendingLinks);
    }
    get aboutLinksArray() {
        return this.parseArrayFromString(this.aboutLinks);
    }
    get siteInformationLinksArray() {
        return this.parseArrayFromString(this.siteInformationLinks);
    }
    get contactNumbersArray() {
        return this.parseArrayFromString(this.contactNumbersList);
    }
    get socialLinksArray() {
        return this.parseArrayFromString(this.socialLinks);
    }
    get isContactcontainerVisible() {
        return !!this.contactNumbersArray.length || !!this.contactUsLink || !!this.chatWithUsLink;
    }
};
__decorate([
    property({ type: String, defaultValue: "" })
], Footer.prototype, "logoUrl", void 0);
__decorate([
    property({ type: String, defaultValue: "[]" })
], Footer.prototype, "quickLinks", void 0);
__decorate([
    property({ type: String, defaultValue: "[]" })
], Footer.prototype, "trendingLinks", void 0);
__decorate([
    property({ type: String, defaultValue: "[]" })
], Footer.prototype, "aboutLinks", void 0);
__decorate([
    property({ type: String, defaultValue: "[]" })
], Footer.prototype, "siteInformationLinks", void 0);
__decorate([
    property({ type: String, defaultValue: "[]" })
], Footer.prototype, "contactNumbersList", void 0);
__decorate([
    property({ type: String, defaultValue: "" })
], Footer.prototype, "localCountryNumbersLink", void 0);
__decorate([
    property({ type: String, defaultValue: "" })
], Footer.prototype, "contactUsLink", void 0);
__decorate([
    property({ type: String, defaultValue: "" })
], Footer.prototype, "chatWithUsLink", void 0);
__decorate([
    property({ type: String, defaultValue: "[]" })
], Footer.prototype, "socialLinks", void 0);
Footer = __decorate([
    customElement({
        tag: "udex-footer",
        renderer: litRender,
        styles: FooterCss,
        template: FooterTemplate,
    })
], Footer);
Footer.define();
export default Footer;
