// @udex/web-components/dist/HeroBanner.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import slot from "@ui5/webcomponents-base/dist/decorators/slot.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import HeroBannerTemplate from "./generated/templates/HeroBannerTemplate.lit.js";
// Styles
import HeroBannerCss from "./generated/themes/HeroBanner.css.js";
var imageLoadingStrategies;
(function (imageLoadingStrategies) {
    imageLoadingStrategies["eager"] = "eager";
    imageLoadingStrategies["lazy"] = "lazy";
})(imageLoadingStrategies || (imageLoadingStrategies = {}));
/**
 * @class
 *
 * @constructor
 * @alias udex.components.HeroBanner
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-hero-banner
 * @public
 */
let HeroBanner = class HeroBanner extends UI5Element {
    get hasBackgroundImage() {
        return !!this.backgroundImage;
    }
    get wrapperBackgroundColor() {
        return this.backgroundColor ? this.backgroundColor : "transparent";
    }
    get hasAdditionalContent() {
        return !!this.additionalContent?.length;
    }
    get hasCustomBackgroundPicture() {
        return !!this.backgroundPicture?.length;
    }
};
__decorate([
    property({ type: String })
], HeroBanner.prototype, "backgroundImage", void 0);
__decorate([
    property({ type: imageLoadingStrategies, defaultValue: imageLoadingStrategies.eager })
], HeroBanner.prototype, "backgroundImageLoadingStrategy", void 0);
__decorate([
    property({ type: String })
], HeroBanner.prototype, "backgroundImageLabel", void 0);
__decorate([
    property({ type: String })
], HeroBanner.prototype, "backgroundColor", void 0);
__decorate([
    slot()
], HeroBanner.prototype, "content", void 0);
__decorate([
    slot()
], HeroBanner.prototype, "additionalContent", void 0);
__decorate([
    slot()
], HeroBanner.prototype, "backgroundPicture", void 0);
HeroBanner = __decorate([
    customElement({
        tag: "udex-hero-banner",
        renderer: litRender,
        styles: HeroBannerCss,
        template: HeroBannerTemplate,
    })
], HeroBanner);
HeroBanner.define();
export default HeroBanner;
