// @udex/web-components/dist/Link.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import "@ui5/webcomponents/dist/Link.js";
import slot from "@ui5/webcomponents-base/dist/decorators/slot.js";
import LinkTemplate from "./generated/templates/LinkTemplate.lit.js";
// Styles
import LinkCss from "./generated/themes/Link.css.js";
import redirectAttributes from "./utils/redirectAttributes.js";
var IconPlacement;
(function (IconPlacement) {
    IconPlacement["start"] = "start";
    IconPlacement["end"] = "end";
})(IconPlacement || (IconPlacement = {}));
/**
 * @class
 *
 * <h3 class="comment-api-title">Overview</h3>
 *
 *
 * <h3>Usage</h3>
 *
 * For the <code>udex-link</code>
 * <h3>ES6 Module Import</h3>
 *
 * <code>import @udex/web-components/dist/Link.js";</code>
 *
 * @constructor
 * @author SAP SE
 * @alias sap.ui.webc.web-components.Link
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-link
 * @public
 */
let Link = class Link extends UI5Element {
    get startIcon() {
        return this.icon && this.iconPlacement === IconPlacement.start;
    }
    get endIcon() {
        return this.icon && this.iconPlacement === IconPlacement.end;
    }
    get linkClass() {
        return this.icon ? "udex-link udex-link--with-icon" : "udex-link";
    }
    onAfterRendering() {
        redirectAttributes(this, "udex-link");
    }
};
__decorate([
    property({ type: String })
], Link.prototype, "icon", void 0);
__decorate([
    property({ type: IconPlacement, defaultValue: IconPlacement.end })
], Link.prototype, "iconPlacement", void 0);
__decorate([
    slot({ type: Node })
], Link.prototype, "default", void 0);
Link = __decorate([
    customElement({
        tag: "udex-link",
        renderer: litRender,
        styles: LinkCss,
        template: LinkTemplate,
        dependencies: [],
    })
], Link);
Link.define();
export default Link;
