// @udex/web-components/dist/Modal.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import slot from "@ui5/webcomponents-base/dist/decorators/slot.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import "@ui5/webcomponents/dist/Dialog.js";
// Template
import ModalTemplate from "./generated/templates/ModalTemplate.lit.js";
// Styles
import ModalCss from "./generated/themes/Modal.css.js";
import redirectAttributes from "./utils/redirectAttributes.js";
var ModalType;
(function (ModalType) {
    ModalType["default"] = "default";
    ModalType["basic"] = "basic";
})(ModalType || (ModalType = {}));
/**
 * @class
 *
 * <h3 class="comment-api-title">Overview</h3>
 *
 * The <code>udex-modal</code> component is a demo component that displays some text.
 *
 * @constructor
 * @alias modal.components.Modal
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-modal
 * @public
 */
let Modal = class Modal extends UI5Element {
    get headerSlot() {
        return !!this.header.length;
    }
    get footerSlot() {
        return !!this.footer.length;
    }
    onAfterRendering() {
        redirectAttributes(this, "modal");
    }
};
__decorate([
    slot({ type: HTMLElement, "default": true })
], Modal.prototype, "content", void 0);
__decorate([
    slot()
], Modal.prototype, "header", void 0);
__decorate([
    slot()
], Modal.prototype, "footer", void 0);
__decorate([
    property({ defaultValue: ModalType.default })
], Modal.prototype, "type", void 0);
__decorate([
    property({ type: Boolean })
], Modal.prototype, "open", void 0);
Modal = __decorate([
    customElement({
        tag: "udex-modal",
        renderer: litRender,
        styles: ModalCss,
        template: ModalTemplate,
    })
], Modal);
Modal.define();
export default Modal;
