// @udex/web-components/dist/Quote.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import slot from "@ui5/webcomponents-base/dist/decorators/slot.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
// Template
import QuoteTemplate from "./generated/templates/QuoteTemplate.lit.js";
// Styles
import QuoteCss from "./generated/themes/Quote.css.js";
var SizeType;
(function (SizeType) {
    SizeType["small"] = "small";
    SizeType["large"] = "large";
})(SizeType || (SizeType = {}));
/**
 * @class
 *
 * <h3 class="comment-api-title">Overview</h3>
 *
 * The <code>udex-quote</code> component is a demo component that displays some text.
 *
 * @constructor
 * @alias modal.components.Quote
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-quote
 * @public
 */
let Quote = class Quote extends UI5Element {
};
__decorate([
    slot({ type: HTMLElement, "default": true })
], Quote.prototype, "content", void 0);
__decorate([
    slot()
], Quote.prototype, "link", void 0);
__decorate([
    slot()
], Quote.prototype, "author", void 0);
__decorate([
    slot()
], Quote.prototype, "image", void 0);
__decorate([
    property({ type: SizeType, defaultValue: "large" })
], Quote.prototype, "imageSize", void 0);
__decorate([
    property({ type: SizeType, defaultValue: "large" })
], Quote.prototype, "textSize", void 0);
Quote = __decorate([
    customElement({
        tag: "udex-quote",
        renderer: litRender,
        styles: QuoteCss,
        template: QuoteTemplate,
    })
], Quote);
Quote.define();
export default Quote;
