// @udex/web-components/dist/RadioButton.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import "@ui5/webcomponents/dist/RadioButton.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import RadioButtonTemplate from "./generated/templates/RadioButtonTemplate.lit.js";
// Styles
import RadioButtonCss from "./generated/themes/RadioButton.css.js";
import redirectAttributes from "./utils/redirectAttributes.js";
/**
 * @class
 *
 * <h3 class="comment-api-title">Overview</h3>
 *
 *
 * <h3>Usage</h3>
 *
 * For the <code>udex-radio-button</code>
 * <h3>ES6 Module Import</h3>
 *
 * <code>import "@udex/web-components/dist/RadioButton.js";</code>
 *
 * @constructor
 * @extends UI5Element
 * @public
 */
let RadioButton = class RadioButton extends UI5Element {
    onAfterRendering() {
        redirectAttributes(this, "udex-radio-button");
    }
    get getReadonly() {
        return !!this.displayOnly;
    }
};
__decorate([
    property({ defaultValue: false, type: Boolean })
], RadioButton.prototype, "displayOnly", void 0);
RadioButton = __decorate([
    customElement({
        tag: "udex-radio-button",
        renderer: litRender,
        styles: RadioButtonCss,
        template: RadioButtonTemplate,
        dependencies: [],
    })
], RadioButton);
RadioButton.define();
export default RadioButton;
