// @udex/web-components/dist/Tab.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var Tab_1;
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import slot from "@ui5/webcomponents-base/dist/decorators/slot.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import executeTemplate from "@ui5/webcomponents-base/dist/renderer/executeTemplate.js";
import TabNavigation from "./TabNavigation.js";
import TabTemplate from "./generated/templates/TabTemplate.lit.js";
import TabHeaderTemplate from "./generated/templates/TabHeaderTemplate.lit.js";
import TabInOverflowTemplate from "./generated/templates/TabInOverflowTemplate.lit.js";
// Styles
import TabInOverflowCss from "./generated/themes/TabInOverflow.css.js";
var Target;
(function (Target) {
    Target["Blank"] = "_blank";
    Target["Parent"] = "_parent";
    Target["Self"] = "_self";
    Target["Top"] = "_top";
})(Target || (Target = {}));
/**
 * @class
 *
 * <h3 class="comment-api-title">Overview</h3>
 *
 *
 * <h3>Usage</h3>
 *
 * For the <code>udex-tab</code>
 * <h3>ES6 Module Import</h3>
 *
 * <code>import @udex/web-components/dist/Tab.js";</code>
 *
 * @constructor
 * @author SAP SE
 * @alias sap.ui.webc.web-components.Tab
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-tab
 * @public
 */
let Tab = Tab_1 = class Tab extends UI5Element {
    getTabInHeaderDomRef() {
        if (this.getElementInHeader) {
            return this.getElementInHeader();
        }
        return null;
    }
    get isOnSelectedTab() {
        return this._realTab === this;
    }
    get _effectiveSlotName() {
        return this.isOnSelectedTab ? this._individualSlot : `disabled-${this._individualSlot}`;
    }
    get hasIcon() {
        return !!this.icon;
    }
    get classWithIcon() {
        return this.hasIcon ? "udex-tab-navigation__button-text--icon" : "";
    }
    get hasNotIcon() {
        return !this.icon && this.showAttentionBadge;
    }
    get hasNotHref() {
        return !this.href;
    }
    get isSelected() {
        return this.selected || this._selected;
    }
    get selectedClass() {
        return this.isSelected ? "udex-tab-navigation__button-tab--active" : "";
    }
    get tabNavigation() {
        return executeTemplate(Tab_1.headerTemplate, this);
    }
    get tabInOverflow() {
        return executeTemplate(Tab_1.overflowTemplate, this);
    }
    get _slotName() {
        return this.isOnSelectedTab ? this._individualSlot : `disabled-${this._individualSlot}`;
    }
    get textValidated() {
        return this.text.length > 25 ? this.text.slice(0, 25) : this.text;
    }
    get effectiveSelected() {
        return this.selected || this._selected;
    }
    get ariaLabelledBy() {
        const labels = [];
        if (this.text) {
            labels.push(`${this.__id}-text`);
        }
        if (this.icon) {
            labels.push(`${this.__id}-icon`);
        }
        return labels.join(" ");
    }
    static get headerTemplate() {
        return TabHeaderTemplate;
    }
    static get overflowTemplate() {
        return TabInOverflowTemplate;
    }
};
__decorate([
    property({ type: String })
], Tab.prototype, "text", void 0);
__decorate([
    property({ type: Boolean })
], Tab.prototype, "selected", void 0);
__decorate([
    property({ type: Boolean })
], Tab.prototype, "showAttentionBadge", void 0);
__decorate([
    property({ type: String })
], Tab.prototype, "icon", void 0);
__decorate([
    property({ type: String })
], Tab.prototype, "href", void 0);
__decorate([
    property({ type: Target, defaultValue: "_self" })
], Tab.prototype, "target", void 0);
__decorate([
    property({ type: Object })
], Tab.prototype, "_realTab", void 0);
__decorate([
    property({ type: Boolean })
], Tab.prototype, "_selected", void 0);
__decorate([
    property({ type: String })
], Tab.prototype, "_tabIndex", void 0);
__decorate([
    slot({ type: Node, "default": true })
], Tab.prototype, "content", void 0);
Tab = Tab_1 = __decorate([
    customElement({
        tag: "udex-tab",
        renderer: litRender,
        template: TabTemplate,
        dependencies: [],
    })
], Tab);
Tab.define();
TabNavigation.registerStaticAreaTabStyles(TabInOverflowCss);
export default Tab;
