// @udex/web-components/dist/TextField.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import Integer from "@ui5/webcomponents-base/dist/types/Integer.js";
// Template
import TextFieldTemplate from "./generated/templates/TextFieldTemplate.lit";
// Styles
import TextFieldCss from "./generated/themes/TextField.css.js";
var ValueState;
(function (ValueState) {
    ValueState["Standard"] = "Standard";
    ValueState["Warning"] = "Warning";
    ValueState["Error"] = "Error";
    ValueState["Success"] = "Success";
})(ValueState || (ValueState = {}));
/**
 * @class
 *
 * @constructor
 * @alias udex.components.TextField
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-text-field
 * @public
 */
let TextField = class TextField extends UI5Element {
    constructor() {
        super(...arguments);
        this._labelActiveClassName = "udex-text-field__label--active";
        this.handleInput = (event) => {
            this.value = event.target?.value;
        };
        this.handleFocus = () => {
            if (!this.disabled && !this.readonly && !this.displayonly) {
                this.labelActiveClass = this._labelActiveClassName;
            }
        };
        this.handleBlur = () => {
            if (this.value) {
                this.labelActiveClass = this._labelActiveClassName;
            }
            else {
                this.labelActiveClass = "";
            }
        };
        this.handleKeyDown = (event) => {
            event.stopPropagation();
        };
    }
    onBeforeRendering() {
        if (this.value && this.labelActiveClass === "") {
            this.labelActiveClass = this._labelActiveClassName;
        }
        if (this.direction === "left") {
            this.dirValue = "ltr";
        }
        else if (this.direction === "right") {
            this.dirValue = "rtl";
        }
        this.isInputReadonly = this.readonly || this.displayonly;
        this.labelClasses = `udex-text-field__label ${this.labelActiveClass}`;
    }
};
__decorate([
    property({ type: Boolean })
], TextField.prototype, "disabled", void 0);
__decorate([
    property({ type: String })
], TextField.prototype, "value", void 0);
__decorate([
    property({ validator: Integer })
], TextField.prototype, "maxlength", void 0);
__decorate([
    property({ type: String })
], TextField.prototype, "label", void 0);
__decorate([
    property({ type: Boolean })
], TextField.prototype, "readonly", void 0);
__decorate([
    property({ type: Boolean })
], TextField.prototype, "displayonly", void 0);
__decorate([
    property({ type: Boolean })
], TextField.prototype, "required", void 0);
__decorate([
    property({ type: ValueState, defaultValue: "Standard" })
], TextField.prototype, "valueState", void 0);
__decorate([
    property({ type: String, defaultValue: "left" })
], TextField.prototype, "direction", void 0);
__decorate([
    property({ type: String })
], TextField.prototype, "supportingText", void 0);
__decorate([
    property({ type: String, noAttribute: true, defaultValue: "" })
], TextField.prototype, "labelActiveClass", void 0);
__decorate([
    property({ type: String, noAttribute: true, defaultValue: "ltr" })
], TextField.prototype, "dirValue", void 0);
__decorate([
    property({ type: Boolean, noAttribute: true, defaultValue: false })
], TextField.prototype, "isInputReadonly", void 0);
__decorate([
    property({ type: String, noAttribute: true, defaultValue: "" })
], TextField.prototype, "labelClasses", void 0);
TextField = __decorate([
    customElement({
        tag: "udex-text-field",
        renderer: litRender,
        styles: TextFieldCss,
        template: TextFieldTemplate,
    })
], TextField);
TextField.define();
export default TextField;
