// @udex/web-components/src/Footer.ts
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";

import "@ui5/webcomponents-icons/dist/headset.js";
import "@ui5/webcomponents-icons/dist/email.js";
import "@ui5/webcomponents-icons/dist/back-to-top.js";
import "@ui5/webcomponents-icons/dist/discussion-2.js";
import "@ui5/webcomponents-icons/dist/navigation-right-arrow.js";

// Template
import FooterTemplate from "./generated/templates/FooterTemplate.lit.js";

// Styles
import FooterCss from "./generated/themes/Footer.css";

type FooterLink = {
  title: string;
  url: string;
};

type FooterSocialLink = {
  url: string;
  iconUrl: string;
};

type FooterContactNumber = {
  number: string;
  country: string;
};

@customElement({
  tag: "udex-footer",
  renderer: litRender,
  styles: FooterCss,
  template: FooterTemplate,
})
class Footer extends UI5Element {
  /**
   * Defines the Logo URL
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.logoUrl
   * @defaultvalue ""
   * @public
   */
  @property({ type: String, defaultValue: "" })
    logoUrl!: string;

  /**
   * List of the Quick links
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.quickLinks
   * @defaultvalue "[]"
   * @public
   */
  @property({ type: String, defaultValue: "[]" })
    quickLinks!: string;

  /**
   * List of the Trending links
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.trendingLinks
   * @defaultvalue "[]"
   * @public
   */
  @property({ type: String, defaultValue: "[]" })
    trendingLinks!: string;

  /**
   * List of the About SAP links
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.aboutLinks
   * @defaultvalue "[]"
   * @public
   */
  @property({ type: String, defaultValue: "[]" })
    aboutLinks!: string;

  /**
   * List of the Site Information links
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.siteInformationLinks
   * @defaultvalue "[]"
   * @public
   */
  @property({ type: String, defaultValue: "[]" })
    siteInformationLinks!: string;

  /**
   * List of the contact numbers
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.contactNumbersList
   * @defaultvalue "[]"
   * @public
   */
  @property({ type: String, defaultValue: "[]" })
    contactNumbersList!: string;

  /**
   * Defines the local country numbers link's href
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.localCountryNumbersLink
   * @defaultvalue ""
   * @public
   */
  @property({ type: String, defaultValue: "" })
    localCountryNumbersLink!: string;

  /**
   * Defines the Contact us link's href
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.contactUsLink
   * @defaultvalue ""
   * @public
   */
  @property({ type: String, defaultValue: "" })
    contactUsLink!: string;

  /**
   * Defines the Chat with us link's href
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.chatWithUsLink
   * @defaultvalue ""
   * @public
   */
  @property({ type: String, defaultValue: "" })
    chatWithUsLink!: string;

  /**
   * List of the Social links and their icons
   * @type {String}
   * @name sap.ui.webc.udex.Footer.prototype.socialLinks
   * @defaultvalue "[]"
   * @public
   */
  @property({ type: String, defaultValue: "[]" })
    socialLinks!: string;

  onBackToTopClick(): void {
    window.scrollTo(0, 0);
  }

  expandLinksList(event: MouseEvent): void {
    let currentElement = event.target as HTMLElement | null;

    while (
      currentElement
      && !currentElement.classList.contains("udex-footer__links-group-container")
    ) {
      currentElement = currentElement.parentElement;
    }

    if (currentElement) {
      currentElement.classList.toggle("active");
    }
  }

  parseArrayFromString<T>(string: string): T[] {
    return string ? JSON.parse(string) as T[] : [];
  }

  get quickLinksArray(): FooterLink[] {
    return this.parseArrayFromString<FooterLink>(this.quickLinks);
  }

  get trendingLinksArray(): FooterLink[] {
    return this.parseArrayFromString<FooterLink>(this.trendingLinks);
  }

  get aboutLinksArray(): FooterLink[] {
    return this.parseArrayFromString<FooterLink>(this.aboutLinks);
  }

  get siteInformationLinksArray(): FooterLink[] {
    return this.parseArrayFromString<FooterLink>(this.siteInformationLinks);
  }

  get contactNumbersArray(): FooterContactNumber[] {
    return this.parseArrayFromString<FooterContactNumber>(this.contactNumbersList);
  }

  get socialLinksArray(): FooterSocialLink[] {
    return this.parseArrayFromString<FooterSocialLink>(this.socialLinks);
  }

  get isContactcontainerVisible(): boolean {
    return !!this.contactNumbersArray.length || !!this.contactUsLink || !!this.chatWithUsLink;
  }
}

Footer.define();

export default Footer;
