// @udex/web-components/src/HeroBanner.ts
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import slot from "@ui5/webcomponents-base/dist/decorators/slot.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";

import HeroBannerTemplate from "./generated/templates/HeroBannerTemplate.lit.js";

// Styles
import HeroBannerCss from "./generated/themes/HeroBanner.css.js";

enum imageLoadingStrategies {
  eager = "eager",
  lazy = "lazy"
}

/**
 * @class
 *
 * @constructor
 * @alias udex.components.HeroBanner
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-hero-banner
 * @public
 */
@customElement({
  tag: "udex-hero-banner",
  renderer: litRender,
  styles: HeroBannerCss,
  template: HeroBannerTemplate,
})
class HeroBanner extends UI5Element {
  /**
   * Allows control over the background image source.
   *
   * @type {string}
   * @name sap.ui.webc.web-components.HeroBanner.prototype.backgroundImage
   * @public
   */
  @property({ type: String })
    backgroundImage?: string;

  /**
   * Allows control over the background image loading strategy.
   *
   * @type {string}
   * @name sap.ui.webc.web-components.HeroBanner.prototype.backgroundImage
   * @public
   */
  @property({ type: imageLoadingStrategies, defaultValue: imageLoadingStrategies.eager })
    backgroundImageLoadingStrategy?: imageLoadingStrategies;

  /**
   * Used as the label for background image.
   *
   * @type {string}
   * @name sap.ui.webc.web-components.HeroBanner.prototype.backgroundImageLabel
   * @public
   */
  @property({ type: String })
    backgroundImageLabel?: string;

  /**
   * Allows control over the background color.
   * Value provided should be compatible with standard CSS rendering, e.g:
   *  - HEX: "#FFF", "#FFFFFF", "FFFFFFFF"
   *  - RGB/A: "rgb(255, 255, 255)", "rgba(255, 255, 255, 1)"
   *  - HSL/A: "hsl(0, 100%, 100%)", "hsla(0, 100%, 100%, 1"
   *  - Variables: "var(--sapShell_Background)", "var(--udexColorNeutralWhite)"
   *  - Text values: "white", "red"
   *
   * @type {string}
   * @name sap.ui.webc.web-components.HeroBanner.prototype.backgroundColor
   * @public
   */
  @property({ type: String })
    backgroundColor?: string;

  /**
  * Slot for primary content.
  *
  * @type {HTMLElement[]}
  * @name sap.ui.webc.web-components.HeroBanner.prototype.content
  * @slot
  * @public
  */
  @slot()
    content?: Array<HTMLElement>;

  /**
  * Slot for any additional content.
  *
  * @type {HTMLElement[]}
  * @name sap.ui.webc.web-components.HeroBanner.prototype.additionalContent
  * @slot
  * @public
  */
  @slot()
    additionalContent?: Array<HTMLElement>;

  /**
  * Slot background picture override.
  *
  * @type {HTMLElement[]}
  * @name sap.ui.webc.web-components.HeroBanner.prototype.content
  * @slot
  * @public
  */
  @slot()
    backgroundPicture?: Array<HTMLElement>;

  get hasBackgroundImage() {
    return !!this.backgroundImage;
  }

  get wrapperBackgroundColor() {
    return this.backgroundColor ? this.backgroundColor : "transparent";
  }

  get hasAdditionalContent() {
    return !!this.additionalContent?.length;
  }

  get hasCustomBackgroundPicture() {
    return !!this.backgroundPicture?.length;
  }
}

HeroBanner.define();

export default HeroBanner;
