// @udex/web-components/src/ProgressBar.ts
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";
import "@ui5/webcomponents/dist/ProgressIndicator.js";

import ProgressBarTemplate from "./generated/templates/ProgressBarTemplate.lit.js";

// Styles
import ProgressBarCss from "./generated/themes/ProgressBar.css.js";
import redirectAttributes from "./utils/redirectAttributes.js";

/**
 * @class
 *
 * <h3 class="comment-api-title">Overview</h3>
 *
 *
 * <h3>Usage</h3>
 *
 * For the <code>udex-progress-bar</code>
 * <h3>ES6 Module Import</h3>
 *
 * <code>import "@udex/web-components/dist/ProgressBar.js";</code>
 *
 * @constructor
 * @author SAP SE
 * @alias sap.ui.webc.web-components.ProgressBar
 * @extends sap.ui.webc.base.UI5Element
 * @tagname udex-progress-bar
 * @public
 */
@customElement({
  tag: "udex-progress-bar",
  renderer: litRender,
  styles: ProgressBarCss,
  template: ProgressBarTemplate,
  dependencies: [],
})
class ProgressBar extends UI5Element {
  onAfterRendering() {
    redirectAttributes(this, "udex-progress-bar");
  }
}

ProgressBar.define();

export default ProgressBar;
