// @udex/web-components/src/RadioButton.ts
import UI5Element from "@ui5/webcomponents-base/dist/UI5Element.js";
import customElement from "@ui5/webcomponents-base/dist/decorators/customElement.js";
import litRender from "@ui5/webcomponents-base/dist/renderer/LitRenderer.js";

import "@ui5/webcomponents/dist/RadioButton.js";
import property from "@ui5/webcomponents-base/dist/decorators/property.js";
import RadioButtonTemplate from "./generated/templates/RadioButtonTemplate.lit.js";

// Styles
import RadioButtonCss from "./generated/themes/RadioButton.css.js";
import redirectAttributes from "./utils/redirectAttributes.js";

/**
 * @class
 *
 * <h3 class="comment-api-title">Overview</h3>
 *
 *
 * <h3>Usage</h3>
 *
 * For the <code>udex-radio-button</code>
 * <h3>ES6 Module Import</h3>
 *
 * <code>import "@udex/web-components/dist/RadioButton.js";</code>
 *
 * @constructor
 * @extends UI5Element
 * @public
 */
@customElement({
  tag: "udex-radio-button",
  renderer: litRender,
  styles: RadioButtonCss,
  template: RadioButtonTemplate,
  dependencies: [],
})
class RadioButton extends UI5Element {
    /**
     * Define if arrow icons for last and first page are visible.
     * @name udex.components.RadioButton.prototype.displayOnly
     * @public
     * @type { boolean }
     * @defaultValue false
     */
    @property({ defaultValue: false, type: Boolean })
      displayOnly!: boolean;

    onAfterRendering() {
      redirectAttributes(this, "udex-radio-button");
    }

    get getReadonly() {
      return !!this.displayOnly;
    }
}

RadioButton.define();

export default RadioButton;
